amazons3/amazons3_utils.bal

```diff
@@ -21,158 +20,332 @@ import ballerina/encoding;
+                           map<string>? queryParams = ()) returns error? {
 
-    string canonicalRequest = "";
+    string canonicalRequest = httpVerb;
     string canonicalQueryString = "";
-    string stringToSign = "";
-    string payloadBuilder = "";
-    string authHeader = "";
-    string amzDateStr = "";
-    string shortDateStr = "";
-    string signedHeader = "";
-    string canonicalHeaders = "";
-    string signedHeaders = "";
     string requestPayload = "";
-    string encodedrequestURIValue = "";
-    string signValue = "";
+    map<string> requestHeaders = headers;
+
+    // Generate date strings and put it in the headers map to generate the signature.
+    (string, string)(amzDateStr, shortDateStr) = generateDateString();
+    requestHeaders[X_AMZ_DATE] = amzDateStr;
+
+    // Get canonical URI.
+    string canonicalURI = getCanonicalURI(requestURI);
+
+        // Generate canonical query string.
+    if (queryParams is map<string> && queryParams.length() > 0) {
+        canonicalQueryString = generateCanonicalQueryString(queryParams);
+    }
+
+    // Encode request payload.
+    if (payload == UNSIGNED_PAYLOAD) {
+        requestPayload = payload;
+    } else {
+        requestPayload = encoding:encodeHex(crypto:hashSha256(payload.toByteArray(UTF_8))).toLower();
+        requestHeaders[CONTENT_TYPE] = request.getHeader(CONTENT_TYPE.toLower());
+    }
 
+    // Generete canonical and signed headers.
+    (string, string) (canonicalHeaders,signedHeaders) = generateCanonicalHeaders(headers, request);
+
+    // Generate canonical request.
+    canonicalRequest = string `${canonicalRequest}\n${canonicalURI}\n${canonicalQueryString}\n`;
+    canonicalRequest = string `${canonicalRequest}${canonicalHeaders}\n${signedHeaders}\n${requestPayload}`;
+
+    // Generate string to sign.
+    string stringToSign = generateStringToSign(amzDateStr, shortDateStr,region, canonicalRequest);
+    
+    // Construct authorization signature string.
+    string authHeader = constructAuthSignature(accessKeyId, secretAccessKey, shortDateStr, region, signedHeaders, 
+                            stringToSign);
+    // Set authorization header.
+    request.setHeader(AUTHORIZATION, authHeader);
+}
+
+function setResponseError(string errorMessage) returns error {
+    error err = error(AMAZONS3_ERROR_CODE, { message : errorMessage });
+    return err;
+}
+
+# Funtion to generate the date strings.
+# 
+# + return - amzDate string and short date string.
+function generateDateString() returns (string, string) {
     time:Time|error time = time:toTimeZone(time:currentTime(), "GMT");
+    string amzDateStr;
+    string shortDateStr;
     if (time is time:Time) {
         string|error amzDate = time:format(time, ISO8601_BASIC_DATE_FORMAT);
         string|error shortDate = time:format(time, SHORT_DATE_FORMAT);
         if (amzDate is string) {
             amzDateStr = amzDate;
+
         } else {
-            return amzDate;
+            panic amzDate;
```

Comment:
If we are just panicking here, without changing the error, we can just use `checkpanic`.

```ballerina
            string amzDate = checkpanic time:format(time, ISO8601_BASIC_DATE_FORMAT);
```

If `time:format(time, ISO8601_BASIC_DATE_FORMAT)` returns `error`, it would result in a panic here.